### PR TITLE
Custom gradient macro proof of concept

### PR DESCRIPTION
**Pull request by mohamed82008**
This PR implements a proof of concept for an `@grad` macro that enables users to define custom gradients for some functions in a user-friendly syntax, identical to Tracker's syntax. Please let me know if I made some obvious mistakes, this is my first contribution to ReverseDiff :)

Here is a usage example that is possible with this PR:
```julia
julia> using ReverseDiff, LinearAlgebra
[ Info: Precompiling ReverseDiff [37e2e3b7-166d-5795-8a7a-e32c996b4267]

julia> f(x) = dot(x, x)
f (generic function with 1 method)

julia> f(x::ReverseDiff.TrackedVector) = ReverseDiff.track(f, x)
f (generic function with 2 methods)

julia> ReverseDiff.@grad function f(x)
           println("The custom gradient has been used.")
           xv = ReverseDiff.value(x)
           return dot(xv, xv), ∇ -> (∇ * 2 * xv,)
       end

julia> x = rand(3);

julia> ReverseDiff.gradient(x -> dot(x, x), x)
3-element Array{Float64,1}:
 1.5585250070434835
 1.8206637829549495
 0.32560088933194864

julia> ReverseDiff.gradient(f, x)
The custom gradient has been used.
3-element Array{Float64,1}:
 1.5585250070434835
 1.8206637829549495
 0.32560088933194864
```

### CODE REVIEW

**Comment by mohamed82008** [comment]
I will add some tests once the general approach is approved.

**Comment by AStupidBear** [comment]
@mohamed82008 I want to do `turing_inference` (DiffEqBayes.jl) with ReverseDiff backend because it's considered the fastest. I need to define custom gradient for `concrete_solve`

```
function DiffEqBase.concrete_solve(prob::DiffEqBase.DEProblem, alg::DiffEqBase.DEAlgorithm, u0, p::ReverseDiff.TrackedArray, args...; kwargs...)
    ReverseDiff.track(concrete_solve, prob, alg, u0, p, args...; kwargs...)
end

function DiffEqBase.concrete_solve(prob::DiffEqBase.DEProblem, alg::DiffEqBase.DEAlgorithm, u0, p::Array{<:ReverseDiff.TrackedReal}, args...; kwargs...)
    # need to defined ReverseDiff.collect just like Tracker.collect
    ReverseDiff.track(concrete_solve, prob, alg, ReverseDiff.collect(u0), ReverseDiff.collect(u0), args...; kwargs...)
end

ReverseDiff.@grad function concrete_solve(prob, alg, u0, p::ReverseDiff.TrackedArray, args...; kwargs...)
    y, back = DiffEqBase._concrete_solve_adjoint(prob, alg, nothing, ReverseDiff.value(u0), ReverseDiff.value(p), args...; kwargs...)
    return y, back
    isnothing(save_idxs) && return y, back
    y[save_idxs, :], function (Δ)
        Δ′ = zero(ReverseDiff.value(y))
        Δ′[save_idxs, :] .= ReverseDiff.value(Δ)
        return back(Δ′)
    end
end
```
However, this PR still cannot handle keyword arguments.

**Comment by mohamed82008** [comment]
@AStupidBear ReverseDiff doesn't seem very actively developed so I have the macro in https://github.com/TuringLang/DistributionsAD.jl/pull/58 hopefully temporarily. I will try to add kwarg support before merging.

File: src/macros.jl [hunk 1]
```diff
@@ -166,3 +166,92 @@ end
...
+
+The `@grad` macro provides a way for the users to define custom adjoints for single-output functions wrt to their input numbers or arrays.
+"""
+macro grad(expr)
+    if @capture(expr, 
+        (f_(xs__) where {T__} = body_) | 
+        (f_(xs__) = body_) | 
+        (function f_(xs__) body_ end) | 
+        (function f_(xs__) where {T__} body_ end)
+    )
```
Comment (oxinabox) [inline src/macros.jl:186]: For breaking appart function bodies I suggest using ExprTools.jl's `splitdef`
there are a bunch of edge cases around this, turns out there are a lot of AST possibilities for how a function can be declared, and it handles all of them, robustly.

Comment (mohamed82008) [reply src/macros.jl:186]: Ya in DistributionsAD, I ended up using MacroTools to support kwargs as well.

**Comment by mohamed82008** [comment]
I believe this is ready for a second review.

File: src/macros.jl [hunk 2]
```diff
@@ -166,3 +166,125 @@ end
...
+                output, input = instruction.output, instruction.input
+                $ReverseDiff.pull_value!.(input)
+                pullback = instruction.cache[2]
+                kwargs = instruction.cache[3]
+                out_value = pullback(input...; kwargs...)[1]
+                $ReverseDiff.value!(output, out_value)
+                return nothing
+            end
+        end
+    end |> esc
```
Comment (ChrisRackauckas) [inline src/macros.jl:238]: is this necessary?

Comment (mohamed82008) [reply src/macros.jl:238]: Well I guess I can just `esc` the closure, argument types and where parameters but I just used `gensym` and `esc` all the way.

Comment (ChrisRackauckas) [reply src/macros.jl:238]: oh yes, it's safe with gensym anyways, I see.

**Comment by ChrisRackauckas** [comment]
Let's get a check from @oxinabox as well.

**Comment by mohamed82008** [comment]
@oxinabox can I merge this?

**Comment by mohamed82008** [comment]
Ping @oxinabox

**Comment by mohamed82008** [comment]
Given the lack of response and given that the tests are passing, I will merge now and release. This is an independent feature so if there are bugs, I can fix them in patch releases later. I need this for other developments in TuringLang, so sorry for my lack of patience.

**Comment by ChrisRackauckas** [comment]
Sounds good to me. No worries!

**Pull request merged**